src/app/CriticalSectionImplRtos.cpp

```diff
@@ -19,17 +19,22 @@
 #include "CriticalSection.h"
 #include "FreeRTOS.h"
 #include "task.h"
-
+#include <pthread.h>
 namespace chip {
 namespace app {
+static pthread_mutex_t eventLocker = PTHREAD_MUTEX_INITIALIZER;
 void CriticalSectionEnter()
 {
+    int err = pthread_mutex_lock(&eventLocker);
+    assert(err == 0);
```

Comment:
Should not use pthread_mutex_lock, since that is not available on many embedded (e.g. most embedded toolchains do not do pthread on FreeRTOS). Furthermore, CriticalSectionEnter is a reentrant call, and it is expected that on FreeRTOS, it also blocks any interrupts that can touch tasks.

The mutex here would have to be re-entrant and it still wouldn't do the right thing.

The right way for this change is for the correct `taskENTER_CRITICAL`/`taskEXIT_CRITICAL` be used. 

Perhaps it should be set using setters before startup of CHIP event loop, so that each platform provides these primitves, rather than relying on an Rtos common code path.